march_hardware/src/ethercat/ethercat_master.cpp

```diff
@@ -185,6 +189,16 @@ void EthercatMaster::ethercatLoop()
       total_loops = 0;
       not_achieved_count = 0;
     }
+
+    const auto delta_t = std::chrono::high_resolution_clock::now() - this->valid_slaves_timestamp_ms_;
+    const auto slave_lost_duration = std::chrono::duration_cast<std::chrono::milliseconds>(delta_t);
+    const std::chrono::milliseconds slave_watchdog_timeout(this->slave_watchdog_timeout_);
+
+    if (slave_lost_duration > slave_watchdog_timeout)
+    {
+      throw error::HardwareException(error::ErrorType::SLAVE_LOST_TIMOUT, "Slave connection lost for %i ms",
```

Comment:
I think there is a problem here. Since `ethercatLoop` runs in its own thread I am not entirely sure how this exception will be caught and if it will be possible to see this exception or not. We might have to use something like [`std::exception_ptr`](https://en.cppreference.com/w/cpp/error/exception_ptr) to propagate the exception across threads. Anyway, it would be important that the ethercat loop stops running and notifies the controller thread that his has happened.

Reply:
The error was visible after the given time so that is kind off nice but it seemed indeed like the hardware interface did not immediately stopped running. I'll call u to see if we can produce a proper solution.